sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SQLHadoopMapReduceCommitProtocol.scala

```diff
@@ -63,7 +70,29 @@ class SQLHadoopMapReduceCommitProtocol(
         committer = ctor.newInstance()
       }
     }
+    totalPartitions = new AtomicInteger(0)
     logInfo(s"Using output committer class ${committer.getClass.getCanonicalName}")
     committer
   }
+
+  override def newTaskTempFile(
+      taskContext: TaskAttemptContext, dir: Option[String], ext: String): String = {
+    val path = super.newTaskTempFile(taskContext, dir, ext)
+    totalPartitions.incrementAndGet()
+    if (dynamicPartitionOverwrite) {
+      if (totalPartitions.get > maxDynamicPartitions) {
```

Comment:
If my understanding on SQLConf.DYNAMIC_PARTITION_MAX_PARTITIONS is correct, that it is the max number of partitions a data source can have at given time, then I am not convinced that this is the right place to check the total number of partitions in the data source. The best would have been [here](https://github.com/apache/spark/blob/76ebf2241a3f2149de13d6c89adcb86325b06004/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala#L137) 
But it might be too late by the time control reaches there.
One way is we can pass the initial partitions in the data source at the beginning, and here can check if the combined number is more than the configured value. But I am not sure if this is the safest way (can another insert happen in parallel on the data source and commit, this affecting the value known here.)

Reply:
> it is the max number of partitions a data source can have at any given time

I don't think so. We should keep the same syntax with Hive. Notice that **hive.exec.max.dynamic.partitions** (default value being 1000) is the total number of dynamic partitions could be created by one DML. https://cwiki.apache.org/confluence/display/Hive/Tutorial#Tutorial-Dynamic-PartitionInsert

Reply:
But you remind me to update it in the SQL documentation.

Reply:
> > it is the max number of partitions a data source can have at any given time
> 
> I don't think so. We should keep the same syntax with Hive. Notice that in Hive's documentation: https://cwiki.apache.org/confluence/display/Hive/Tutorial#Tutorial-Dynamic-PartitionInsert
> 
> > hive.exec.max.dynamic.partitions (default value being 1000) is the total number of dynamic partitions could be created **by one DML**.

I saw the two links that you shared in this PR. [One](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DML#LanguageManualDML-DynamicPartitionInserts), that you had specified in the Description of this PR. and Another [here](https://cwiki.apache.org/confluence/display/Hive/Tutorial#Tutorial-Dynamic-PartitionInsert). It is not clear to me if `hive.exec.max.dynamic.partitions` is per DML or for the life-time of the data source.. Please clarify.

Reply:
Thank you for point it. The descriptions are both from official documentations, but I prefer the [second one](https://cwiki.apache.org/confluence/display/Hive/Tutorial#Tutorial-Dynamic-PartitionInsert). Actually, there is no restriction for life-time of the data source in Hive, unless it stored into Hive Metastore. But these parameters hive.exec.max.dynamic.partitions.* only take effect in Hive Client. So they are for per DML.

Reply:
Here is a simple testing for Hive table:
```sql
bin/spark-sql --master local --hiveconf hive.exec.dynamic.partition.mode=nonstrict --hiveconf hive.exec.dynamic.partition=true --hiveconf hive.exec.max.dynamic.partitions=3

spark-sql>create temporary view t4 (key, value) as select 2, explode(array(1, 2, 3, 4));
spark-sql>select * from t4;
2	1
2	2
2	3
2	4
>insert into table p partition(key) select key, value from t4;
Error in query: org.apache.hadoop.hive.ql.metadata.HiveException: Number of dynamic partitions created is 4, which is more than 3. To solve this try to set hive.exec.max.dynamic.partitions to at least 4.;

spark-sql>create temporary view t4 (key, value) as select 2, explode(array(1, 2, 3));
spark-sql>select * from t3;
2	1
2	2
2	3
spark-sql>insert into table p partition(key) select key, value from t3;
spark-sql>insert into table p partition(key) select key, value from t3;
spark-sql>select * from p;
2	1
2	1
2	2
2	2
2	3
2	3
```

Reply:
oh thanks. now it makes things clearer. Thanks.